src/cascadia/TerminalSettingsModel/Profile.cpp

```diff
@@ -533,5 +539,141 @@ Json::Value Profile::ToJson() const
     JsonUtils::SetValueForKey(json, BellStyleKey, _BellStyle);
     JsonUtils::SetValueForKey(json, PixelShaderPathKey, _PixelShaderPath);
 
+    JsonUtils::SetValueForKey(json, EnvironmentKey, _EnvironmentVariables);
+
     return json;
 }
+
+namespace
+{
+    class VariablePair
+    {
+    public:
+        VariablePair(std::wstring rawValueIn) :
+            rawValue(std::move(rawValueIn))
+        {
+        }
+        VariablePair() = default;
+        virtual ~VariablePair() = default;
+        std::wstring rawValue;
+        std::wstring resolvedValue;
+    };
+
+    std::wstring ResolveEnvironmentVariableValue(const std::wstring& key, std::map<std::wstring, VariablePair>& envMap, std::list<std::wstring> seenKeys = {})
+    {
+        auto it = envMap.find(key);
+        if (it != envMap.end())
+        {
+            if (!it->second.resolvedValue.empty())
+            {
+                return it->second.resolvedValue;
+            }
+            if (std::find(seenKeys.cbegin(), seenKeys.cend(), key) != seenKeys.end())
+            {
+                std::wstringstream error;
+                error << L"Self referencing keys in environment settings: ";
+                for (auto seenKey : seenKeys)
+                {
+                    error << L"'" << seenKey << L"', ";
+                }
+                error << L"'" << key << L"'";
+                throw winrt::hresult_error(WEB_E_INVALID_JSON_STRING, winrt::hstring(error.str()));
+            }
+            const std::wregex parameterRegex(L"\\$\\{env:(.*?)\\}");

```

Comment:
this could be static, right? It's always going to be the same in every call of this function.

Also, you might want to use a [raw string literal](https://stackoverflow.com/questions/56710024/what-is-a-raw-string) for the regex. That _should_ result in less escaping. I think it should be:

```c++
const std::wregex parameterRegex(R"(\$\{env:(.*?)\})");
```

but I might have that slightly wrong